plugins/catalog-backend/src/next/database/ProcessingDatabaseImpl.ts

```diff
@@ -0,0 +1,244 @@
+      .where('entity_id', id);
+
+    if (refreshResult === 0) {
+      throw new NotFoundError(`Processing state not found for ${id}`);
+    }
+
+    // Schedule all deferred entities for future processing.
+    await this.addUnprocessedEntities(tx, {
+      entities: deferedEntities,
+      id,
+      type: 'entity',
+    });
+
+    // Update fragments
+
+    // Delete old relations
+    await tx<DbRelationsRow>('relations')
+      .where({ originating_entity_id: id })
+      .delete();
+
+    // Batch insert new relations
+    const relationRows: DbRelationsRow[] = relations.map(
+      ({ source, target, type }) => ({
+        originating_entity_id: id,
+        source_entity_ref: stringifyEntityRef(source),
+        target_entity_ref: stringifyEntityRef(target),
+        type,
+      }),
+    );
+    await tx.batchInsert(
+      'relations',
+      this.deduplicateRelations(relationRows),
+      BATCH_SIZE,
+    );
+  }
+
+  private deduplicateRelations(rows: DbRelationsRow[]): DbRelationsRow[] {
+    return lodash.uniqBy(
+      rows,
+      r => `${r.source_entity_ref}:${r.target_entity_ref}:${r.type}`,
+    );
+  }
+
+  async addUnprocessedEntities(
+    txOpaque: Transaction,
+    options: AddUnprocessedEntitiesOptions,
+  ): Promise<void> {
+    const tx = txOpaque as Knex.Transaction;
+    const entityIds = new Array<string>();
+
+    for (const entity of options.entities) {
+      const entityRef = stringifyEntityRef(entity);
+      await tx<DbRefreshStateRow>('refresh_state')
+        .insert({
+          entity_id: uuid(),
+          entity_ref: entityRef,
+          unprocessed_entity: JSON.stringify(entity),
+          errors: '',
+          next_update_at: tx.fn.now(),
+          last_discovery_at: tx.fn.now(),
+        })
+        .onConflict('entity_ref')
+        .merge(['unprocessed_entity', 'last_discovery_at']);
+
+      const [{ entity_id: entityId }] = await tx<DbRefreshStateRow>(
+        'refresh_state',
+      ).where({ entity_ref: entityRef });
+      entityIds.push(entityId);
+    }
+
+    const key =
+      options.type === 'provider'
+        ? { source_special_key: options.id }
+        : { source_entity_id: options.id };
+    // copied from update refs
+    await tx<DbRefreshStateReferences>('refresh_state_references')
+      .where(key)
+      .delete();
+
+    const referenceRows: DbRefreshStateReferences[] = entityIds.map(
+      entityId => ({
+        ...key,
+        target_entity_id: entityId,
+      }),
+    );
+    await tx.batchInsert('refresh_state_references', referenceRows, BATCH_SIZE);
+  }
+
+  async getProcessableEntities(
+    txOpaque: Transaction,
+    request: { processBatchSize: number },
+  ): Promise<GetProcessableEntitiesResult> {
+    const tx = txOpaque as Knex.Transaction;
+
+    const items = await tx<DbRefreshStateRow>('refresh_state')
+      .select()
+      .where('next_update_at', '<=', tx.fn.now())
+      .limit(request.processBatchSize)
+      .orderBy('next_update_at', 'desc');
```

Comment:
Wouldn't it be `asc` to get oldest first? 🤔

Reply:
Yeah that seems more correct :grin:

Reply:
Good find Tim! Fixed now 👍

Reply:
Haha, good eye!